### ISSUE TITLE
Defun grad_func should accept the outputs of the function being differentiated

### ISSUE THREAD

**Issue by rizar** [tensorflow/tensorflow#4661]
When dealing with discrete stochastic units, one often wants to define custom gradients. It seems like the most legitimate way to it in Tensorflow is to use `grad_func` keyword argument of `function.Defun` from `tensorflow.python.framework`.  However, the gradient function is only given the inputs of the function defined and the gradients, whereas it would be often very convenient to also use the outputs. 

For example, see the code below.

```python
# In current Tensorflow I have to do like this:

@function.Defun(*(3 * [tf.float32]))
def oneovertwo_unit_bprop(probs, noise, grads):
  outcome = tf.to_float(tf.less(noise, probs))
  outcome_probs = outcome * probs + (1 - outcome) * (1 - probs)
  return grads / 2. / outcome_probs, tf.zeros_like(noise)


@function.Defun(*(2 * [tf.float32]), grad_func=oneovertwo_unit_bprop)
def oneovertwo_unit_fprop(probs, noise):
  return tf.to_float(tf.less(noise, probs))


def oneovertwo_unit(probs):
  return oneovertwo_unit_fprop(probs, tf.random_uniform(tf.shape(probs)))

# I would like to do like this

def oneovertwo_unit_bprop(probs, outcome, grads):
  outcome_probs = outcome * probs + (1 - outcome) * (1 - probs)
  return grads / 2. / outcome_probs


@function.Defun(*(2 * [tf.float32]), grad_func=oneovertwo_unit_bprop)
def oneovertwo_unit(probs):
  return tf.to_float(tf.less(tf.random_uniform(probs.get_shape()), probs))
```

FYI, this an implementation of the gradient estimator from the DARN paper (https://arxiv.org/pdf/1310.8499v2.pdf).

**Comment by concretevitamin** [comment]
@zffchen78: is this something on the radar?  

If not, we can mark it as contributions welcome, as a workaround seems to exist.

**Comment by zffchen78** [comment]
It's work as intended. The seemingly duplicated computation done by tf.to_float(tf.less(noise, probs)) can be eliminated by TF runtime. See example in function_test.py. Look for usage of _OptimizationOptions().

The reasoning behind my design choice is that if the grad_func's signature always gives the implementation the result of the forward function. That typically means the result is needed to be kept around until backprop. It's not always a good choice in face of limited memory.

Instead, if you always recompute the forward compution inside the backward function, TF can have choice to either eliminate the duplicated computation during the whole graph optimization but likely to use more memory; or TF can choose to redo the computation instead using more memory.

In your case, pulling out the random_uniform as the noise is the right choice. Because random_uniform is a stateful operation which outputs different results on each innovation.

BTW, I recently improved the Defun syntax so that you no longer needs specify the type in the Defun decorator constructor. So you can remove *(3 * [tf.float32]) and *(2 * [tf.float32]): E.g.,

```
@function.Defun()
def oneovertwo_unit(probs):
  
  @function.Defun()
  def bprop(probs, noise, grads):
    outcome = tf.to_float(tf.less(noise, probs))
    outcome_probs = outcome * probs + (1 - outcome) * (1 - probs)
    return grads / 2. / outcome_probs, tf.zeros_like(noise)


  @function.Defun(grad_func=bprop)
  def fprop(probs, noise):
    return tf.to_float(tf.less(noise, probs))

  return fprop(probs, tf.random_uniform(tf.shape(probs)))

```
